src/format_source.sh

```diff
@@ -38,16 +36,8 @@ blkrun="-l 88"
 # Black test options
 blktest="--check"
 
+cd "$base"
 # Directories to process
-cppdirs="libtoast libtoast_mpi toast"
-pydirs="toast ../pipelines"
-
-# Test
-for cppd in ${cppdirs}; do
-    find "${base}/${cppd}" -name "*.hpp" -not -path '*Random123*' -not -path '*pybind11/*' -not -path '*gtest/*' -exec ${unexe} ${unrun} '{}' \;
-    find "${base}/${cppd}" -name "*.cpp" -not -path '*Random123*' -not -path '*pybind11/*' -not -path '*gtest/*' -exec ${unexe} ${unrun} '{}' \;
-done
-
-for pyd in ${pydirs}; do
-    find "${base}/${pyd}" -name "*.py" -not -path '*pybind11/*' -exec ${blkexe} ${blkrun} '{}' \;
-done
+find libtoast toast \( -name "*.hpp" -or -name "*.cpp" \) -and -not \( -path '*Random123*' -or -path '*pybind11/*' -or -path '*gtest/*' \) -exec ${unexe} ${unrun} {} + &
+find toast ../pipelines -name "*.py" -not -path '*pybind11/*' -exec ${blkexe} ${blkrun} {} + &
```

Comment:
Thanks, I was not aware of this syntax until now.  Note that the `+` argument to `find ... -exec` simply passes the result of the find (the long list of files) to the command.  So it only works with commands that accept multiple files as an argument (which both uncrustify and black do).  I've added a comment to the script in the toast3 branch about this.

Reply:
I checked them before patching it, but in practice, of all command line program accepting path as positional argument, I don't recall any of them not accepting multiple paths, except in those cases that requires exactly 2 paths which has different meaning (input path, output path, such as `git mv`, etc.) Also, program like this often support "recursive", so providing directories instead is enough if not because of those filters used in find (they may have cli args that can do similar filtering but probably using find as a single syntax used to provide filter to 2 completely unrelated program is smarter.)

But one thing I did hesitate in the patch is POSIX compliance! `+` (and the other useful `-print0`) is not required by POSIX, and is a more "modern" addition from GNU find.

If I remember correctly, BSD find didn't support them years ago. But when I create this patch, BSD find on macOS 10.15+ is already supporting it. As the performance implication is big, I think probably it is ok. But I should have mentioned this in the original message first. (The "need for speed" comes from wrestling with the formatter to try to manually edit it in some way that the formatter operation is an identity operation... E.g. black try to split `[:]` in 3 lines and I didn't managed to manually write something else that it would preserve.)

(Writing bash script that's compatible with any system out there is hard.)